Consolidate specialist finder configuration into the schemas

ChrisBAshton opened the pull request:
There's no reason why we need to spread finder configs across both a schema JSON file _and_ an ActiveModel.
This PR migrates the following data out of the models and into the schemas, making it easier to add and edit finders, and reducing room for bugs/discrepancies:

- Taxons
- Primary publishing organisation
- Title
- Rename `slug` to `admin_slug`

Ideas for the future:

- Move `admin_slug` into the schemas. Easier said than done: attempted in 595eaca071e2528f1be9b92c966a2ee6d7313989 and 8cea1b547cfb8fbda209adfc54ef6e6be9c6ad46
- Format specific fields. Easier said than done: attempted in 52a62b87c222a9d36b66936c3ddd8569c981e227. There are a handful of other weird method definitions in the specialist document models which means we can't just blindly delete the initialise method.
- Move the specialist document models into their own directory. Easier said than done: attempted in https://github.com/alphagov/specialist-publisher/commit/dfe52f3f09a18ac2e22a016b859dcc5bdfc1073b

Trello: https://trello.com/c/Bw2sYhwE/2776-consolidate-finder-configs-in-specialist-publisher

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

ChrisBAshton committed:
Move `:editing_organisations` next to `:organisations` keyword arg

This is the order in which they're set as instance variables, so
let's make things consistent.

ChrisBAshton committed:
Refactor unit tests for FinderSchema

- Removes hard dependency on real schema JSON file data
- Makes much clearer which part of the schema is actually being
  tested
- Adds missing tests for the 'getter' methods

ChrisBAshton committed:
Make FinderSchema take a schema (hash) as an argument

This will make it easier to test different branches of logic - see
next commit.

ChrisBAshton committed:
Add `taxons` property to FinderSchema

ChrisBAshton committed:
Have `Document` read `taxons` from the finder schema

This consolidates the `taxons` configuration into the schema,
rather than having to be specified in the model.

ChrisBAshton committed:
Move primary publishing organisation to top of list of orgs

Every other finder lists the `primary_publishing_organisation`
as the first organisation in its array of `organisations`. This
commit makes that consistent for
algorithmic_transparency_records.json too.

ChrisBAshton committed:
Update primary publishing org for BusinessFinanceSupportScheme

This value should match what's in the schema:
https://github.com/alphagov/specialist-publisher/blob/5ddc02c90dad50d3f3e35bcbf0613148554a5d2f/lib/documents/schemas/business_finance_support_schemes.json#L14

The old `2bde479a-97f2-42b5-986a-287a623c2a1c` value should
probably have been removed in bb7567ac0432d087fd45c7d996ec641fbff6eb14.
Nobody appears to have noticed because all of this finder's
documents are `unpublished` anyway:
https://specialist-publisher.publishing.service.gov.uk/business-finance-support-schemes

ChrisBAshton committed:
Remove duplicated 'primary publishing organisation'

Now that all schema JSON files consistently have the primary
publishing org ID as the first organisation in the list of
`organisations`, we can remove the duplicated IDs from the models.

ChrisBAshton committed:
Add missing #facets test

ChrisBAshton committed:
Copy 'title' from models into schemas

In the next commit, we'll use the schema title instead of the
model title, and then we can remove the titles from the models.
This consolidates the data into the one place it should clearly
be!

ChrisBAshton committed:
Retrieve `title` from schema instead of model

In the next commit we'll remove the `title` method from the rest
of the models.

ChrisBAshton committed:
Remove `self.title` - this is now derived from the schema

ryanb-gds commented on app/lib/finder_schema.rb:
```diff
@@ -6,15 +6,21 @@ def self.schema_names
+    JSON.parse(File.read(Rails.root.join("lib/documents/schemas/#{type}.json")))
+  end
+
+  attr_reader :base_path, :organisations, :editing_organisations, :taxons, :format, :content_id, :document_title
```
Might want to consolidate the schema attribute reader to here.

ChrisBAshton replied:
Good idea.

ryanb-gds commented on app/models/document.rb:
```diff
@@ -185,6 +172,10 @@ def facet_options(facet)
   delegate :organisations, to: :finder_schema
 
+  def primary_publishing_organisation
+    organisations.first
```
Is this a safe assumption? Just double checking

ChrisBAshton replied:
It's the truth for all the schema JSON files at the moment - I checked them 👍 only risk is if someone doesn't follow this approach in future.

But worth noting that #1230 introduces a feature for the "Statutory Instruments" finder to override the primary publishing organisation per-document. (Subsequently rolled out to FloodAndCoastalErosionRiskManagementResearchReport and LicenceTransaction too).

ChrisBAshton replied:
Hmmm I'll take another look at that :/

ChrisBAshton replied:
Just deployed the branch to integration and was able to change the primary publishing org on https://specialist-publisher.integration.publishing.service.gov.uk/flood-and-coastal-erosion-risk-management-research-reports/6993f8c2-8936-492c-a5c5-4ea4b7444f97:en - so I think we're ok.

ryanb-gds reviewed (approved):
Good work on this, definitely makes things more consistent. Pretty sure you'll have considered the primary publishing org selection so going to approve this on that assumption!

ChrisBAshton committed:
Rename `slug` to `admin_slug`

This slug is only surfaced in the browser in the Specialist
Publisher admin UI. So renaming to avoid confusion (could
otherwise be referring to the public slug of the finder itself,
or the slug of a specialist document).

ChrisBAshton committed:
Consolidate schema into list of attr_reader
